substrate/frame/support/procedural/src/runtime/parse/pallet_decl.rs

```diff
@@ -42,20 +47,127 @@ impl PalletDeclaration {
 
 		let mut path = path.path.clone();
 
+		let mut pallet_segment = None;
+		let mut runtime_param = None;
 		let mut instance = None;
 		if let Some(segment) = path.segments.iter_mut().find(|seg| !seg.arguments.is_empty()) {
 			if let PathArguments::AngleBracketed(syn::AngleBracketedGenericArguments {
 				args, ..
 			}) = segment.arguments.clone()
 			{
-				if let Some(syn::GenericArgument::Type(syn::Type::Path(arg_path))) = args.first() {
-					instance =
-						Some(Ident::new(&arg_path.to_token_stream().to_string(), arg_path.span()));
+				if segment.ident == "Pallet" {
+					let mut segment = segment.clone();
 					segment.arguments = PathArguments::None;
+					pallet_segment = Some(segment.clone());
+				}
+				let mut args_iter = args.iter();
+				if let Some(syn::GenericArgument::Type(syn::Type::Path(arg_path))) =
+					args_iter.next()
+				{
+					let ident =
+						Ident::new(&arg_path.to_token_stream().to_string(), arg_path.span());
+					if segment.ident == "Pallet" {
+						runtime_param = Some(ident.clone());
+						if let Some(arg_path) = args_iter.next() {
+							instance = Some(Ident::new(
+								&arg_path.to_token_stream().to_string(),
+								arg_path.span(),
+							));
+							segment.arguments = PathArguments::None;
+						}
+					} else {
+						instance = Some(ident);
+						segment.arguments = PathArguments::None;
+					}
 				}
 			}
 		}
 
-		Ok(Self { name, path, instance, attrs: item.attrs.clone() })
+		if pallet_segment.is_some() {
+			path = syn::Path {
+				leading_colon: None,
+				segments: path.segments.first().cloned().into_iter().collect(),
```

Comment:
```suggestion
				segments: path.segments.iter().filter(|seg| seg.arguments.is_empty()).cloned().iter()
```

There is right now no guarantee that `path.segments.len() == 2`